### PR TITLE
Clarify `optional::swap` conditions

### LINKED ISSUE #5033: Strange condition in optional::swap
looks like a typo

```cpp    
void swap(optional& _Right)
        noexcept(...) {
        static_assert(is_move_constructible_v<_Ty>, ...);
        static_assert(!is_move_constructible_v<_Ty> || is_swappable_v<_Ty>, ...)
....

```

### PR DESCRIPTION
**Pull request by KSARK**
resolves #5033

### CODE REVIEW

File: stl/inc/optional [hunk 1]
```diff
@@ -349,10 +349,12 @@ public:
 
     _CONSTEXPR20 void swap(optional& _Right)
         noexcept(is_nothrow_move_constructible_v<_Ty> && is_nothrow_swappable_v<_Ty>) {
-        static_assert(is_move_constructible_v<_Ty>,
-            "optional<T>::swap requires T to be move constructible (N4950 [optional.swap]/1).");
-        static_assert(!is_move_constructible_v<_Ty> || is_swappable_v<_Ty>,
-            "optional<T>::swap requires T to be swappable (N4950 [optional.swap]/2).");
+        if constexpr (is_move_constructible_v<_Ty>) {
+            static_assert(
+                is_swappable_v<_Ty>, "optional<T>::swap requires T to be swappable (N4950 [optional.swap]/2).");
```
Comment (CaseyCarter) [inline stl/inc/optional:354]: We may as well update these working draft citations while we're here (I verified that referenced paragraphs are still correct in the new draft):
```suggestion
                is_swappable_v<_Ty>, "optional<T>::swap requires T to be swappable (N4993 [optional.swap]/2).");
```

File: stl/inc/optional [hunk 2]
```diff
@@ -349,10 +349,12 @@ public:
...
         noexcept(is_nothrow_move_constructible_v<_Ty> && is_nothrow_swappable_v<_Ty>) {
-        static_assert(is_move_constructible_v<_Ty>,
-            "optional<T>::swap requires T to be move constructible (N4950 [optional.swap]/1).");
-        static_assert(!is_move_constructible_v<_Ty> || is_swappable_v<_Ty>,
-            "optional<T>::swap requires T to be swappable (N4950 [optional.swap]/2).");
+        if constexpr (is_move_constructible_v<_Ty>) {
+            static_assert(
+                is_swappable_v<_Ty>, "optional<T>::swap requires T to be swappable (N4950 [optional.swap]/2).");
+        } else {
+            static_assert(false, "optional<T>::swap requires T to be move constructible (N4950 [optional.swap]/1).");
```
Comment (CaseyCarter) [inline stl/inc/optional:356]: Ditto:
```suggestion
            static_assert(false, "optional<T>::swap requires T to be move constructible (N4993 [optional.swap]/1).");
```

**Review by CaseyCarter — APPROVED** [review]
Just a couple of drive-by edits that I'll apply.

**Comment by CaseyCarter** [comment]
Thanks! We merge pull requests in batches, so don't expect to see much activity here until later in the week.

**Comment by StephanTLavavej** [comment]
I'm mirroring this to the MSVC-internal repo - please notify me if any further changes are pushed.

**Comment by StephanTLavavej** [comment]
Thanks for improving these diagnostics, and congratulations on your second microsoft/STL commit! :partying_face: :heart_eyes_cat: :confetti_ball:

This change is expected to ship in VS 2022 17.13 Preview 3.

**Comment by StephanTLavavej** [comment]
I am going to label and changelog this as https://github.com/microsoft/STL/labels/enhancement instead of https://github.com/microsoft/STL/labels/documentation because it affects diagnostics (compiler errors) instead of documentation or code comments.

**Pull request merged**